### ISSUE TITLE
Integrate Potree to ThreeJS

### ISSUE THREAD

**Issue by imwithye** [potree/potree#1380]
I am trying to integrate potree to existing threejs project (without using the offical viewer)

Here is what I have tried

```javascript
/* global Potree */

import { WebGLRenderer, Scene, PerspectiveCamera, BoxGeometry, MeshBasicMaterial, Mesh } from 'three'
import { OrbitControls } from 'three/examples/jsm/controls/OrbitControls.js'

Potree.loadPointCloud('./L1_potree_data/metadata.json', 'T', (e) => {
  const camera = new PerspectiveCamera(45, window.innerWidth / window.innerHeight, 1, 10000)
  const scene = new Scene()
  const renderer = new WebGLRenderer()
  renderer.setSize(window.innerWidth, window.innerHeight)
  document.body.appendChild(renderer.domElement)

  const pointcloud = e.pointcloud
  const pointclouds = [pointcloud]
  scene.add(pointcloud)

  const pointcloudBox = pointcloud.pcoGeometry.tightBoundingBox ? pointcloud.pcoGeometry.tightBoundingBox : pointcloud.boundingBox
  const center = pointcloudBox.min.clone().add(pointcloudBox.max).multiplyScalar(0.5)
  camera.position.copy(center)

  const controls = new OrbitControls(camera, renderer.domElement)
  const testCube = new Mesh(new BoxGeometry(10, 10, 10), new MeshBasicMaterial({ color: 0x00ff00 }))
  scene.add(testCube)

  const loop = () => {
    Potree.updatePointClouds(pointclouds, camera, renderer)
    controls.update()
    renderer.render(scene, camera)
    requestAnimationFrame(loop)
  }
  loop()
})
```

Although the octree.bin is loaded, but there is nothing in the scene. Coud potree provide an official example for ThreeJS?

**Comment by michael-erskine** [comment]
Might I suggest using one of the community ports such as pnext/three-loader (supports v1 clouds) or shiukaheng/potree-loader (supports Potree v2 clouds). Both of these are aimed at direct usage in other general-purpose three.js projects and I have used both successfully.

**Comment by hamza-iqbal-ad** [comment]
> Might I suggest using one of the community ports such as pnext/three-loader (supports v1 clouds) or shiukaheng/potree-loader (supports Potree v2 clouds). Both of these are aimed at direct usage in other general-purpose three.js projects and I have used both successfully.

Yes I have used them but they do not have support for s3 presigned urls. I am up for any suggestion to use presigned urls for these packages

**Comment by victorchrollo14** [comment]
> > Might I suggest using one of the community ports such as pnext/three-loader (supports v1 clouds) or shiukaheng/potree-loader (supports Potree v2 clouds). Both of these are aimed at direct usage in other general-purpose three.js projects and I have used both successfully.
> 
> Yes I have used them but they do not have support for s3 presigned urls. I am up for any suggestion to use presigned urls for these packages

checkout this fork, it has some example implementation for using s3 presigned urls. 
https://github.com/aeromosaico/potree

But the issue now is, that the new potree format ( metadata.json, heirarchy.bin, octree.bin ) doesn't seem to work fine, only the old format with cloud.js and few millions  .bin files works fine, but can't really use presigned urls for such a large number of files.

**Comment by im-hamza-dev** [comment]
In order to use protected models like presigned urls for s3, we have to inject urls somehow into the fetch process of files because three-loader doesn't provide this feature and even we couldn't have generic urls.

Workaround for this is to override the fetch function like this:




```
let urls={
  octree: ".....",
  metadata:"......",
  hierarchy:"......",
}
 insertPresignedUrls(urls) {

    const nativeFetch = window.fetch;

    window.fetch = function (...args) {

      if (args[0].toString().includes('hierarchy.bin')) {

        args[0] = urls.hierarchy;

      } else if (args[0].toString().includes('octree.bin')) {

        args[0] = urls.octree;

      } else if (args[0].toString().includes('metadata.json')) {

        args[0] = urls.metadata;

      }

      return nativeFetch.apply(window, args);

    };

  }
```

**Comment by victorchrollo14** [comment]
> In order to use protected models like presigned urls for s3, we have to inject urls somehow into the fetch process of files because three-loader doesn't provide this feature and even we couldn't have generic urls.
> 
> Workaround for this is to override the fetch function like this:
> 
> 
> 
> 
> ```
> let urls={
>   octree: ".....",
>   metadata:"......",
>   hierarchy:"......",
> }
>  insertPresignedUrls(urls) {
> 
>     const nativeFetch = window.fetch;
> 
>     window.fetch = function (...args) {
> 
>       if (args[0].toString().includes('hierarchy.bin')) {
> 
>         args[0] = urls.hierarchy;
> 
>       } else if (args[0].toString().includes('octree.bin')) {
> 
>         args[0] = urls.octree;
> 
>       } else if (args[0].toString().includes('metadata.json')) {
> 
>         args[0] = urls.metadata;
> 
>       }
> 
>       return nativeFetch.apply(window, args);
> 
>     };
> 
>   }
> ```

Cool.

Btw where u able to use  the new format of potree. The one with the metadata.json file.

**Comment by im-hamza-dev** [comment]
Yes, if you are using pnext/three-loader, you just have to pass v2 in Potree initiator function.

**Comment by victorchrollo14** [comment]
> Yes, if you are using pnext/three-loader, you just have to pass v2 in Potree initiator function.

Thanks 👍. 
Were you able to use it in potree directly?
the new format doesn't seem to work in potree

**Comment by im-hamza-dev** [comment]
@victorchrollo14  yes I have used in Potree directly.
You just have to pass metadata file in first argument, assuming other files are in same folder, it will work.
Below is an example: ( dummy2 folder must have octree.bin and hierarchy.bin files )
`Potree.loadPointCloud("../pointclouds/dummy2/metadata.json", "lion", function(e){}`

**Comment by victorchrollo14** [comment]
> @victorchrollo14 yes I have used in Potree directly. You just have to pass metadata file in first argument, assuming other files are in same folder, it will work. Below is an example: ( dummy2 folder must have octree.bin and hierarchy.bin files ) `Potree.loadPointCloud("../pointclouds/dummy2/metadata.json", "lion", function(e){}`

Yeah I tried this, but it throws a lot of errors. https://github.com/potree/potree/issues/1447

**Comment by im-hamza-dev** [comment]
Sure lets check it there
